internal/downloader/s3.go

```diff
@@ -0,0 +1,49 @@
+package downloader
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/pkg/errors"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+)
+
+type downloaderAPI interface {
+	Download(w io.WriterAt, input *s3.GetObjectInput, options ...func(*s3manager.Downloader)) (n int64, err error)
+	DownloadWithContext(ctx aws.Context, w io.WriterAt, input *s3.GetObjectInput, options ...func(*s3manager.Downloader)) (n int64, err error)
+	DownloadWithIterator(ctx aws.Context, iter s3manager.BatchDownloadIterator, opts ...func(*s3manager.Downloader)) error
+}
+
+type Downloader struct {
+	Bucket string
+	d      downloaderAPI
+}
+
+func New(bucketName string, dl downloaderAPI) (*Downloader, error) {
+	if dl == nil {
+		return nil, fmt.Errorf("received nil downloader")
+	}
```

Comment:
Not sure if this check is required, I have included it because the `s3manager.NewDownloader()` function returns a pointer to a downloader without any error. As a result, I added the `if` statement to handle a potential `nil` pointer case.

Reply:
As it stands today, `NewDownloader` never returns `nil`: https://github.com/aws/aws-sdk-go/blob/master/service/s3/s3manager/download.go#L79

but, the type signature allows it to possibly return nil in the future. I would personally assume that it will never be nil, and simplify this constructor as a result. But I think either way is reasonable enough.

Reply:
I see, I will go ahead and remove the check for now then.